Improve Ollama to implement #summarize and #default_dimension

rickychilcott opened the pull request:
Because Ollama can run multiple models, `embed` will return different embedding vector sizes. Based on the current running model, we will generate an embedding and return the size in `default_dimension` instead of nil. I was not able to find a way to get the size of the embeddings in the docs (see https://github.com/ollama/ollama/blob/main/docs/api.md#show-model-information)

Also, I wanted to implement `summarize` for Ollama, and it seems that there is no dedicated endpoint for that operation. I used the pattern from OpenAI's ChatGPT to use a prompt to get the summary. However, I was consistently getting "Sure, here is a summary of the text" at the beginning of my response, so I modified the prompt specifically for llama2 to give only a summary. I'm not sure it's ideal, because each model will have its own best prompt to return a summary.

In addition, I added specs for `complete`, `summarize`, and `default_dimension`. Since Ollama can chunk output, I couldn't find a way to shim the tests using the existing patterns (something previous devs were seemingly also stumped by). I added VCR which will at least capture the right chunked response and serve it up. However, it will not actually chunk in the same way it does live, so I had to make a minor change to the streaming code to handle it from VCR and live. I don't believe I've broken streaming.

Addresses #380 for Ollama.

rickychilcott commented:
I know https://github.com/andreibondarev/langchainrb/issues/144 was talking about VCR, but it's an older discussion and hand-mocking Faraday calls seems sub-optimal too.

andreibondarev commented on lib/langchain/llm/response/ollama_response.rb:
```diff
@@ -44,5 +44,9 @@ def prompt_tokens
       raw_response.dig("eval_count")
     end
+
+    def summarization
```
I'm not sure we need this method. Since it's basically just a completion based on the prompt instructions -- I think it's okay to just use `.completion`.

andreibondarev commented on lib/langchain/llm/ollama.rb:
```diff
@@ -27,6 +27,13 @@ def initialize(url:, default_options: {})
+    # @return [Integer] The # of vector dimensions
+    def default_dimension
+      # since Ollama can run multiple models, generate an embedding and return the size
+      @default_dimension ||= embed(text: "test").embedding.size
```
What if instead of dynamically calculating it we'd figure out what this size is for every single supported model and store it in a look up hash? 🤔

rickychilcott replied:
I'd be willing to build that hash for a popular few... but all? I'm not a fan of that. If I did a partial list, and then fall back to this method, would that be ok?

andreibondarev replied:
@rickychilcott I'm good with that!

rickychilcott commented:
Ok , I've done this now and merged in `main`.

I checked the embedding size for the four most popular LLMs at https://ollama.com/library, and there definitely is a long tail of models provided by Ollama. I added tests for all and also added a vcr spec for the least popular model (i.e. `tinydolphin`).

andreibondarev commented on lib/langchain/llm/response/base_response.rb:
```diff
@@ -43,6 +43,14 @@ def embedding
+      #
+      # @return [String]
+      #
+      def summarization
```
I think we can remove this method, right?

andreibondarev commented on lib/langchain/llm/ollama.rb:
```diff
@@ -14,7 +14,16 @@ class Ollama < Base
       embeddings_model_name: "llama2",
-      chat_completion_model_name: "llama2"
+      chat_completion_model_name: "llama2",
+      embedding_size: {
```
I'm wondering whether this should exist here or as a separate constant?
```ruby
EMBEDDING_SIZES = {
  codellama: 4_096,
  ...
}
```

rickychilcott replied:
Cool!

andreibondarev commented:
@rickychilcott Thanks for this PR btw! Makes me think that we need to implement this dynamic `default_dimension()` method on other LLMs (in follow-up PRs of course). 

Could you please upmerge this PR?

rickychilcott commented:
Merged main. Thanks @andreibondarev

rickychilcott commented:
@andreibondarev - anything I need to do to get this merged?
